src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs

```diff
@@ -678,11 +678,23 @@ protected override void Shutdown()
             }
             else
             {
-                VariableDetails result =
-                    await editorSession.DebugService.EvaluateExpression(
+                VariableDetailsBase result;
+                if (evaluateParams.Context.Equals("watch", StringComparison.OrdinalIgnoreCase))
+                {
+                    // If we evaluate an expression from a watch context then the server goes into a deadlock.
+                    // So instead, we retrieve the value from already evaluated variables.
+                    result = editorSession.DebugService.GetVariableFromExpression(
```

Comment:
Good point. I didn't think about evaluating an arbitrary expression when I found the underlying issue. I would like to think of this as a temporary fix till we figure out the right way to do it. As I have mentioned in `powershellcontext.cs`, the hang is the result of the server trying to acquire a handle to an already occupied runspace and that runspace won't be release till the script finishes execution. But the bigger question to ask is why does vscode send a `watch` evaluation request when the user `continues` and not when the debugger hits a breakpoint, unlike variables, scopes, etc? Something doesn't seem right here.